cni/network/invoker_cns.go

```diff
@@ -84,78 +86,102 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
 		return IPAMAddResult{}, errors.Wrap(err, "Failed to get IP address from CNS with error: %w")
 	}
 
-	info := IPv4ResultInfo{
-		podIPAddress:       response.PodIpInfo.PodIPConfig.IPAddress,
-		ncSubnetPrefix:     response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
-		ncPrimaryIP:        response.PodIpInfo.NetworkContainerPrimaryIPConfig.IPSubnet.IPAddress,
-		ncGatewayIPAddress: response.PodIpInfo.NetworkContainerPrimaryIPConfig.GatewayIPAddress,
-		hostSubnet:         response.PodIpInfo.HostPrimaryIPInfo.Subnet,
-		hostPrimaryIP:      response.PodIpInfo.HostPrimaryIPInfo.PrimaryIP,
-		hostGateway:        response.PodIpInfo.HostPrimaryIPInfo.Gateway,
-	}
+	addResult := IPAMAddResult{}
 
-	// set the NC Primary IP in options
-	addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
+	for i := 0; i < len(response.PodIpInfo); i++ {
+		info := IPResultInfo{
+			podIPAddress:       response.PodIpInfo[i].PodIPConfig.IPAddress,
+			ncSubnetPrefix:     response.PodIpInfo[i].NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
+			ncPrimaryIP:        response.PodIpInfo[i].NetworkContainerPrimaryIPConfig.IPSubnet.IPAddress,
+			ncGatewayIPAddress: response.PodIpInfo[i].NetworkContainerPrimaryIPConfig.GatewayIPAddress,
+			hostSubnet:         response.PodIpInfo[i].HostPrimaryIPInfo.Subnet,
+			hostPrimaryIP:      response.PodIpInfo[i].HostPrimaryIPInfo.PrimaryIP,
+			hostGateway:        response.PodIpInfo[i].HostPrimaryIPInfo.Gateway,
+		}
 
-	log.Printf("[cni-invoker-cns] Received info %+v for pod %v", info, podInfo)
+		// set the NC Primary IP in options
+		addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
 
-	ncgw := net.ParseIP(info.ncGatewayIPAddress)
-	if ncgw == nil {
-		if invoker.ipamMode != util.V4Overlay {
-			return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
-		}
+		log.Printf("[cni-invoker-cns] Received info %+v for pod %v", info, podInfo)
 
-		ncgw = net.ParseIP(overlayGatewayIP)
-	}
+		ncgw := net.ParseIP(info.ncGatewayIPAddress)
+		if ncgw == nil {
+			if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualModeOverlay) {
```

Comment:
do we need a separate, v6 mode for overlay?

Reply:
The ideal is, to not even have this flag in CNI, and have CNI depend on CNS to decide what kind of ipam mode CNS is using.
For the time being, we are just adding in the extra mode, but want to later refactor this whole thing to not have the flag at all in CNI